service/frontend/authHandler.go

```diff
@@ -0,0 +1,648 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package frontend
+
+import (
+	"context"
+
+	"github.com/uber/cadence/.gen/go/cadence/workflowserviceserver"
+	"github.com/uber/cadence/.gen/go/health"
+	"github.com/uber/cadence/.gen/go/health/metaserver"
+	"github.com/uber/cadence/.gen/go/replicator"
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/auth"
+	"github.com/uber/cadence/common/resource"
+)
+
+// TODO(vancexu): add metrics
+
+const (
+	// ActionCommon is for authority implementation to handle auth for common operations
+	ActionCommon = "common"
+	// ActionAdmin is for authority implementation to handle auth for admin only operations
+	ActionAdmin = "admin"
+)
+
+// AuthHandlerImpl frontend handler wrapper for authentication and authorization
+type AuthHandlerImpl struct {
+	resource.Resource
+
+	frontendHandler workflowserviceserver.Interface
+	authority       auth.Authority
+
+	startFn func()
+	stopFn  func()
+}
+
+var _ workflowserviceserver.Interface = (*AuthHandlerImpl)(nil)
+
+// NewAuthHandlerImpl creates frontend handler with authentication support
+func NewAuthHandlerImpl(wfHandler *DCRedirectionHandlerImpl, authority auth.Authority) *AuthHandlerImpl {
+	if authority == nil {
+		authority = auth.NewNopAuthority()
+	}
+
+	return &AuthHandlerImpl{
+		Resource:        wfHandler.Resource,
+		frontendHandler: wfHandler,
+		authority:       authority,
+		startFn:         func() { wfHandler.Start() },
+		stopFn:          func() { wfHandler.Stop() },
+	}
+}
+
+// TODO(vancexu): refactor frontend handler
+
+// RegisterHandler register this handler, must be called before Start()
+func (a *AuthHandlerImpl) RegisterHandler() {
+	a.GetDispatcher().Register(workflowserviceserver.New(a))
+	a.GetDispatcher().Register(metaserver.New(a))
+}
+
+// Health callback for for health check
+func (a *AuthHandlerImpl) Health(ctx context.Context) (*health.HealthStatus, error) {
+	hs := &health.HealthStatus{Ok: true, Msg: common.StringPtr("auth is good")}
+	return hs, nil
+}
+
+// Start starts the handler
+func (a *AuthHandlerImpl) Start() {
+	a.startFn()
+}
+
+// Stop stops the handler
+func (a *AuthHandlerImpl) Stop() {
+	a.stopFn()
+}
+
+// CountWorkflowExecutions API call
+func (a *AuthHandlerImpl) CountWorkflowExecutions(
+	ctx context.Context,
+	request *shared.CountWorkflowExecutionsRequest,
+) (*shared.CountWorkflowExecutionsResponse, error) {
+
+	isAllowed, err := a.isAllowedForDomain(ctx, request.GetDomain())
+	if err != nil {
+		return nil, err
+	}
+	if !isAllowed {
+		return nil, errUnauthorized
+	}
+	return a.frontendHandler.CountWorkflowExecutions(ctx, request)
+}
+
+// DeprecateDomain API call
+func (a *AuthHandlerImpl) DeprecateDomain(
+	ctx context.Context,
+	request *shared.DeprecateDomainRequest,
+) error {
+
+	isAllowed, err := a.isAdmin(ctx)
```

Comment:
this look weird. How about creating a helper something like 

```
isAuthorized(actor string, domainName string) (bool, error) {}
```
For this api, you can pass the actor as admin